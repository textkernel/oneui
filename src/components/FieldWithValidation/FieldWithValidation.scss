@import '../../themes/oneui/oneui';

.FieldWithValidation {
    margin: var(--space-75) 0 0 0;
}
