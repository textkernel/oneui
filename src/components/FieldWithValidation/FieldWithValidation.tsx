import * as React from 'react';
import { bem } from '../../utils';
import { Tooltip } from '../Tooltip';
import { Text } from '../Text';
import styles from './FieldWithValidation.scss';

export interface Props {
    /** a single form field, e.g. Input, TextArea, etc. It should support context=”danger” prop.
     * When useTooltip is set to true, children that are ReactComponents should support forwardRef */
    children: React.ReactElement;
    /** if this is not empty, this message will be rendered and context danger will be applied to the child */
    errorMessage?: string;
    /** if set to true the error message will be rendered as a tooltip. Otherwise it will be a text under the field */
    useTooltip?: boolean;
}

const { block } = bem('FieldWithValidation', styles);

export const FieldWithValidation: React.FC<Props> = ({
    children,
    errorMessage,
    useTooltip = false,
    ...rest
}) => {
    const [isChildInFocus, setIsChildInFocus] = React.useState(false);

    const handleFocus = () => {
        setIsChildInFocus(true);
    };

    const handleBlur = () => {
        setIsChildInFocus(false);
    };

    if (errorMessage && useTooltip) {
        const clonedChild = errorMessage
            ? React.cloneElement(children, {
                  context: 'danger',
                  onFocus: handleFocus,
                  onBlur: handleBlur,
              })
            : children;

        return (
            <Tooltip {...rest} content={errorMessage} visible={isChildInFocus}>
                {clonedChild}
            </Tooltip>
        );
    }

    if (errorMessage) {
        const clonedChild = React.cloneElement(children, { context: 'danger' });

        return (
            <>
                {clonedChild}
                <Text {...rest} {...block()} context="danger" size="small">
                    {errorMessage}
                </Text>
            </>
        );
    }

    return children;
};

FieldWithValidation.displayName = 'FieldWithValidation';
