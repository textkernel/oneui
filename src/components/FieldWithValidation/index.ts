export { FieldWithValidation, Props as FieldWithValidationProps } from './FieldWithValidation';
