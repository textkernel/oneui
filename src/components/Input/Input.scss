@import '../../themes/oneui/oneui';

.Input {
    @extend %input;

    border-color: var(--color-neutral-20);

    &--context {
        @each $context in ($contexts) {
            &_#{ $context } {
                border-color: var(--color-#{$context}-50);
                &:focus {
                    border-color: var(--color-#{$context}-50);
                }
            }
        }
    }

    &::placeholder {
        color: var(--color-neutral-30);
        opacity: 1;
    }

    &--isBlock {
        width: 100%;
    }

    &[disabled] {
        background-color: var(--color-background);
    }

    &--size {
        &_small {
            min-height: 26px;
            height: 26px;
            font-size: var(--font-size-small);
            line-height: var(--line-height-small);
            padding: var(--space-50);
        }
        &_large {
            font-size: var(--font-size-large);
            line-height: var(--line-height-large);
            padding: var(--space-100);
        }
    }
}
