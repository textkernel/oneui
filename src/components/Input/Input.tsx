import * as React from 'react';
import { bem } from '../../utils';
import styles from './Input.scss';
import { ValidationContext, InputType, Size } from '../../constants';

// Any other attributes (onChange, onKeyUp etc.) are
// supported although not defined in props type definition
export interface Props extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'size'> {
    /** The input field context (e.g. brand, critical, success etc. - defaults to brand) */
    context?: ValidationContext;
    /** Should the input field be disabled or not */
    disabled?: boolean;
    /** Whether or not to show block-level input field (full width) */
    isBlock?: boolean;
    /** The size of the input field */
    size?: Size;
    /** Type of the input field */
    type?: InputType;
}

const { block } = bem('Input', styles);

export const Input = React.forwardRef<HTMLInputElement, Props>(
    (
        {
            children,
            context,
            disabled = false,
            isBlock = false,
            size = 'normal',
            type = 'text',
            value,
            ...rest
        },
        ref
    ) => {
        const isLastPassDisabled = type !== 'password';
        return (
            <input
                {...rest}
                {...block({ context, size, isBlock, disabled, ...rest })}
                ref={ref}
                type={type}
                disabled={disabled}
                value={value}
                data-lpignore={isLastPassDisabled}
            />
        );
    }
);

Input.displayName = 'Input';
