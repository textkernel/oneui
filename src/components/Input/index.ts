export { Input, Props as InputProps } from './Input';
