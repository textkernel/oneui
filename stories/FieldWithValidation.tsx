import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { text, boolean, withKnobs } from '@storybook/addon-knobs';
import { FieldWithValidation, Input, TextArea } from '@textkernel/oneui';

const Example = () => {
    const [inputValue, setValue] = React.useState('');
    const [errMsg, setErrMsg] = React.useState<string>();
    const EMAIL_REGEX =
        // eslint-disable-next-line no-useless-escape
        /^(([^<>()\[\]\.,;:\s@\"]+(\.[^<>()\[\]\.,;:\s@\"]+)*)|(\".+\"))@(([^<>()[\]\.,;:\s@\"]+\.)+[^<>()[\]\.,;:\s@\"]{2,})$/i;

    React.useEffect(() => {
        if (!inputValue) {
            setErrMsg('This field is required');
        } else {
            setErrMsg(
                inputValue.match(EMAIL_REGEX) ? undefined : 'This is not a valid email address'
            );
        }
    }, [inputValue, EMAIL_REGEX]);

    return (
        <div style={{ marginBottom: '10px' }}>
            <FieldWithValidation
                errorMessage={errMsg}
                useTooltip={boolean('Use tooltip for input field', false)}
            >
                <Input
                    value={inputValue}
                    onChange={(e) => {
                        const { value } = e.target;
                        setValue(value);
                    }}
                />
            </FieldWithValidation>
        </div>
    );
};

storiesOf('Molecules/FieldWithValidation', module)
    .addDecorator(withKnobs)
    .add('FieldWithValidation', () => (
        <>
            <div style={{ marginBottom: '10px' }}>
                <FieldWithValidation
                    errorMessage={text('Error message for input field', 'Something wrong')}
                    useTooltip={boolean('Use tooltip for input field', false)}
                >
                    <Input />
                </FieldWithValidation>
            </div>
            <div>
                <FieldWithValidation
                    errorMessage={text('Error message for text field', '')}
                    useTooltip={boolean('Use tooltip for text field', false)}
                >
                    <TextArea />
                </FieldWithValidation>
            </div>
        </>
    ))
    .add('Example implementation', () => <Example />);
