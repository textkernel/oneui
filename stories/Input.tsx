import * as React from 'react';
import { Input } from '@textkernel/oneui';

export default {
    title: 'Atoms/Input',
    component: Input,
};

export const Uncontrolled = (args) => <Input {...args} />;
Uncontrolled.args = {
    placeholder: 'Some text goes here...',
};

export const Controlled = (args) => <Input {...args} />;
Controlled.args = {
    placeholder: 'While typing, check your console log...',
    onChange: (e) => {
        console.log(e.target.value);
    },
    value: '',
};
