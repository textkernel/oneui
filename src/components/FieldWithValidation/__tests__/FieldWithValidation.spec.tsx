import * as React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { FieldWithValidation } from '../FieldWithValidation';
import { Input } from '../../Input';
import '@testing-library/jest-dom';

describe('FieldWithValidation', () => {
    let wrapper;

    describe('when rendering message as text', () => {
        describe('with no error message passed', () => {
            beforeEach(() => {
                wrapper = render(
                    <FieldWithValidation>
                        <Input />
                    </FieldWithValidation>
                );
            });

            it('should simply render the child as it is', () => {
                expect(wrapper.asFragment()).toMatchSnapshot();
                expect(screen.getByRole('textbox')).toBeInTheDocument();
            });
        });
    });

    describe('with error message passed', () => {
        const message = 'invalid field';
        beforeEach(() => {
            wrapper = render(
                <FieldWithValidation errorMessage={message}>
                    <Input />
                </FieldWithValidation>
            );
        });

        it('should set context to bad on child', () => {
            expect(wrapper.asFragment()).toMatchSnapshot();
            const input = screen.getByRole('textbox');
            expect(input).toHaveAttribute('class', 'Input Input--context_bad');
            expect(input).toBeInTheDocument();
        });

        it('should render the message as text', () => {
            expect(screen.queryByText(message)).toBeDefined();
        });
    });
    describe('when using tooltip', () => {
        describe('with no error message passed', () => {
            beforeEach(() => {
                wrapper = render(
                    <FieldWithValidation useTooltip>
                        <Input />
                    </FieldWithValidation>
                );
            });

            it('should not render a tooltip with no error message', () => {
                expect(wrapper.asFragment()).toMatchSnapshot();
                fireEvent.change(screen.getByDisplayValue(''), { target: { value: 'Utrecht' } });
            });
        });
        describe('when error message is defined', () => {
            const message = 'invalid field';
            beforeEach(() => {
                wrapper = render(
                    <FieldWithValidation errorMessage={message} useTooltip>
                        <Input />
                    </FieldWithValidation>
                );
            });

            it('should set context to bad on child', () => {
                expect(wrapper.asFragment()).toMatchSnapshot();
                const input = screen.getByRole('textbox');
                expect(input).toHaveAttribute('class', 'Input Input--context_bad');
            });
            it('should render the message when field is focused', () => {
                expect(wrapper.asFragment()).toMatchSnapshot();
                expect(screen.queryByText(message)).toBeDefined();
            });
            it('should remove the message when field is blurred', () => {
                expect(wrapper.asFragment()).toMatchSnapshot();
                expect(screen.queryByText(message)).toBeDefined();
            });
        });
    });
});
