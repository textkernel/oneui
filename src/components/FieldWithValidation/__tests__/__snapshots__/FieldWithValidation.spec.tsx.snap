// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldWithValidation when rendering message as text with no error message passed should simply render the child as it is 1`] = `
<DocumentFragment>
  <input
    class="Input"
    data-lpignore="true"
    type="text"
    value=""
  />
</DocumentFragment>
`;

exports[`FieldWithValidation when using tooltip when error message is defined should remove the message when field is blurred 1`] = `
<DocumentFragment>
  <input
    class="Input Input--context_bad"
    data-lpignore="true"
    type="text"
    value=""
  />
</DocumentFragment>
`;

exports[`FieldWithValidation when using tooltip when error message is defined should render the message when field is focused 1`] = `
<DocumentFragment>
  <input
    class="Input Input--context_bad"
    data-lpignore="true"
    type="text"
    value=""
  />
</DocumentFragment>
`;

exports[`FieldWithValidation when using tooltip when error message is defined should set context to bad on child 1`] = `
<DocumentFragment>
  <input
    class="Input Input--context_bad"
    data-lpignore="true"
    type="text"
    value=""
  />
</DocumentFragment>
`;

exports[`FieldWithValidation when using tooltip with no error message passed should not render a tooltip with no error message 1`] = `
<DocumentFragment>
  <input
    class="Input"
    data-lpignore="true"
    type="text"
    value=""
  />
</DocumentFragment>
`;

exports[`FieldWithValidation with error message passed should set context to bad on child 1`] = `
<DocumentFragment>
  <input
    class="Input Input--context_bad"
    data-lpignore="true"
    type="text"
    value=""
  />
  <p
    class="Text--context_bad Text--size_small FieldWithValidation"
  >
    invalid field
  </p>
</DocumentFragment>
`;
